server/src/main/java/io/deephaven/server/session/SessionServiceGrpcImpl.java

```diff
@@ -177,13 +167,45 @@ public void exportFromTicket(
 
         final SessionState.ExportObject<Object> source = ticketRouter.resolve(
                 session, request.getSourceId(), "sourceId");
-        session.newExport(request.getResultId(), "resultId")
+        final SessionState.ExportObject<Object> destination = session.newExport(request.getResultId(), "resultId")
                 .require(source)
                 .onError(responseObserver)
-                .submit(() -> {
-                    GrpcUtil.safelyComplete(responseObserver, ExportResponse.getDefaultInstance());
-                    return source.get();
+                .submit(source::get);
+        session.nonExport()
+                .require(destination)
+                .onError(responseObserver)
+                .submit(() -> GrpcUtil.safelyComplete(responseObserver, ExportResponse.getDefaultInstance()));
+    }
+
+    @Override
+    public void publishFromTicket(
+            @NotNull final PublishRequest request,
+            @NotNull final StreamObserver<PublishResponse> responseObserver) {
+        final SessionState session = service.getCurrentSession();
+
+        if (!request.hasSourceId()) {
+            responseObserver
+                    .onError(Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT, "Source ticket not supplied"));
+            return;
+        }
+        if (!request.hasResultId()) {
+            responseObserver
+                    .onError(Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT, "Result ticket not supplied"));
+            return;
+        }
+
+        final SessionState.ExportObject<Object> source = ticketRouter.resolve(
+                session, request.getSourceId(), "sourceId");
+        Ticket resultId = request.getResultId();
+
+        final SessionState.ExportBuilder<Object> publisher = ticketRouter.publish(
+                session, resultId, "resultId", () -> {
+                    // when publish is complete, complete the gRPC request
+                    GrpcUtil.safelyComplete(responseObserver, PublishResponse.getDefaultInstance());
                 });
+        publisher.require(source)
+                .onError(responseObserver)
+                .submit(source::get);
```

Comment:
I'm still not completely in-tune w/ all the export builder constructions. Colin should approve

Reply:
We must not complete the gRPC until publisher was complete, since it is possible that the client could get the response and start using the new ticket before it was actually written to.

publisher.require(source).submit(source::get) ensures that source exists and is ready for use, reads the underlying object out of source, passes it to the publisher. If we completed after calling source.get(), the client in theory could receive the message before the value was written into resultId. Then the runnable passed to ticketRouter.publish() will run after the publish iis complete, ensuring that the value is ready for use (but _not_ guaranteeing that someone else doesn't stomp on it in the meantime...).

So this one seems correct to me at least, and I don't think I would have any changes to make here.